distributed/worker_state_machine.py

```diff
@@ -2077,109 +2096,146 @@ def _transition_resumed_released(
+            if ts.done:
+                return {ts: "waiting"}, []
+            else:
+                ts.state = "resumed"
+                ts.next = "waiting"
+                return {}, []
 
     def _transition_cancelled_forgotten(
         self, ts: TaskState, *, stimulus_id: str
     ) -> RecsInstrs:
-        ts.next = "forgotten"
-        if not ts.done:
-            return {}, []
         return {ts: "released"}, []
 
     def _transition_cancelled_released(
         self, ts: TaskState, *, stimulus_id: str
     ) -> RecsInstrs:
         if not ts.done:
             return {}, []
-        self.executing.discard(ts)
-        self.long_running.discard(ts)
-        self.in_flight_tasks.discard(ts)
 
-        self._release_resources(ts)
-        return self._transition_generic_released(ts, stimulus_id=stimulus_id)
+        if ts.previous in ("executing", "long-running"):
+            # Calls _ensure_computing
+            recs, instr = self._transition_from_executing(ts)
+        else:
+            assert ts.previous == "flight"
+            # No need to call _ensure_communicating, since it's done at the end of
+            # every transition loop
+            recs, instr = {}, []
+
+        return merge_recs_instructions(
+            (recs, instr),
+            self._transition_generic_released(ts, stimulus_id=stimulus_id),
+        )
 
     def _transition_executing_released(
         self, ts: TaskState, *, stimulus_id: str
     ) -> RecsInstrs:
-        ts.previous = ts.state
+        """We can't stop executing a task just because the scheduler asked us to,
+        so we're entering cancelled state and waiting until it completes.
+        """
+        assert ts.state in ("executing", "long-running")
+        ts.previous = cast(Literal["executing", "long-running"], ts.state)
         ts.next = None
-        # See https://github.com/dask/distributed/pull/5046#discussion_r685093940
         ts.state = "cancelled"
         ts.done = False
         return {}, []
 
-    def _transition_generic_memory(
+    def _transition_executing_memory(
         self, ts: TaskState, value: object = NO_VALUE, *, stimulus_id: str
     ) -> RecsInstrs:
-        if value is NO_VALUE and ts.key not in self.data:
-            raise RuntimeError(
-                f"Tried to transition task {ts} to `memory` without data available"
-            )
+        if self.validate:
+            assert value is not NO_VALUE or ts.key in self.data
+
+        recs1, instr1 = self._transition_from_executing(ts)
 
-        instructions: Instructions = []
         try:
-            recs = self._put_key_in_memory(ts, value, stimulus_id=stimulus_id)
+            recs2 = self._put_key_in_memory(ts, value, stimulus_id=stimulus_id)
         except Exception as e:
+            ts.state = "released"
             msg = error_message(e)
-            recs = {ts: tuple(msg.values())}
+            recs2 = {ts: tuple(msg.values())}
+            instr2: Instructions = []
```

Comment:
IIUC this would create a released->error transition which is not defined.

By not setting `ts.state = "released"` we'd get an ordinary `executing->error` transition. The serialization _should_ be picked up as a task compute / user error which I think is OK.

Reply:
`released->error` is defined, and we need it anyway in case the serialization problem is caused by scatter (although I'd bet the use case is untested).